middle_end/flambda/flambda_iterators.mli

```diff
@@ -35,6 +35,8 @@ val map_subexpressions
   -> Flambda.t
   -> Flambda.t
 
+val map_tail_subexpressions : (Flambda.t -> Flambda.t) -> Flambda.t -> Flambda.t
```

Comment:
This function counts regions, exclaves and the body of `try` as 'tail'. This is fine, but please document it here: not all definitions of 'tail' in the compiler agree on this.

Reply:
I suggest changing the name of this function; it seems potentially confusing otherwise.

Reply:
What would be a better name? AFAICT, the only term ever used elsewhere for “tail” in this sense is “tail” (with the aforementioned inconsistency). I was going to call it `map_region_tail_subexpressions` but that sounds like it's going to map over the tail of the expression's region (implying it will do something clever with `exclave` and `region` expressions).

The problem is, for _nearly_ all purposes after the frontend, the definition used here the relevant one. We can state it as “a subexpression in this position pops the stack” (both stacks, that is). In particular, these are the subexpressions that (a) are _allowed_ to pop explicitly and (b) will _implicitly_ pop if they don't. That puts the body of a `try` block squarely in the “tail” camp! (Even though they actually only ever implicitly pop the stack.)

The problem is we have this annoying property of function calls, that their semantics depends on whether they're in “tail position” in a much more narrowly useful sense. That seems like something we should be fixing on the way out of source code? If I could tell by looking at an application whether it clobbers the stack frame, would we even need the other concept of tail position in the middle end?

(Also, now that I'm thinking about it, we should just have exclaves for the control stack as well. Then we get rid of `reg_close` and we have _exactly one semantics for function calls_.)